### PR TITLE
ADef_attack added

### PR DESCRIPTION
**Pull request by EvgeniaAR**
Please comment on further required changes, e.g. considering the tests, I am not sure whether further ones are required.

Thank you
Evgenia

### CODE REVIEW

**Comment by wielandbrendel** [comment]
@EvgeniaAR : looks pretty good! The coverage for your attacks is currently 93%. Could you add tests for the missing line or mark them with `# pragma: no cover` to avoid that they are counted against the coverage statistic if tests don't make any sense?

File: foolbox/attacks/adef_attack.py [hunk 1]
```diff
@@ -0,0 +1,398 @@
...
+        return new_image
+
+
+def _create_vec_field(fval, gradf, d1x, d2x, color_axis, smooth=0):
+    """Calculate the deformation vector field
+
+    In:
+    fval: float
+    gradf: numpy.ndarray
+        of shape Cxhxw with C=3 or C=1 (color channels), h,w >= 1.
```
Comment (wielandbrendel) [inline foolbox/attacks/adef_attack.py]: Here you talk about color_axis as 3 or 1 but later use python indexing, i.e. 0 or 2. Could you align this?

**Comment by jonasrauber** [comment]
I am still reviewing (familiarizing myself with the paper and the attack), but Wieland already raised some important new points. In addition here are two more: 1) whether the attack is targeted or not should depend on the criterion (whether it provides a target class or not) and not on a keyword argument of the attack and 2) could you please fix the formatting in the docstrings and comments, e.g. spacing after commas, etc. :)

File: foolbox/attacks/adef_attack.py (hunk 1 above)
Comment (jonasrauber) [reply foolbox/attacks/adef_attack.py]: @wielandbrendel not sure which other location you are referring to, but for the shape here, only 1 and 3 make sense, so that's correct (@EvgeniaAR but spaces around the equal sign and after the comma are missing :P )

Comment (wielandbrendel) [reply foolbox/attacks/adef_attack.py]: Sorry, you are right, I misread that part (I thought 1, 3 refers to channel axis, not the number of channels).

File: foolbox/attacks/adef_attack.py [hunk 2]
```diff
@@ -0,0 +1,405 @@
...
+    vec_field[:, :, 0] = -fval * alpha1 / norm_squared_alpha
+    vec_field[:, :, 1] = -fval * alpha2 / norm_squared_alpha
+
+    return vec_field
+
+
+class ADefAttack(Attack):
+    """Adversarial Attack that distorts the image, i.e. changes the locations
+    of pixels. The algorithm is described in [1],
+    A Repository with the original code can be found in [2].
```
Comment (jonasrauber) [inline foolbox/attacks/adef_attack.py]: should be lowercase A

File: foolbox/attacks/adef_attack.py [hunk 3]
```diff
@@ -0,0 +1,405 @@
...
+
+    References
+    ----------
+    .. [1] Rima Alaifari, Giovanni S. Alberti, and Tandri Gauksson1:
+           "ADef: an Iterative Algorithm to Construct Adversarial
+           Deformations", https://arxiv.org/pdf/1804.07729.pdf
```
Comment (jonasrauber) [inline foolbox/attacks/adef_attack.py]: please replace the URL with the non-pdf one: https://arxiv.org/abs/1804.07729

File: foolbox/attacks/adef_attack.py [hunk 4]
```diff
@@ -0,0 +1,405 @@
...
+    vec_field = np.empty((h, w, 2))
+    vec_field[:, :, 0] = -fval * alpha1 / norm_squared_alpha
+    vec_field[:, :, 1] = -fval * alpha2 / norm_squared_alpha
+
+    return vec_field
+
+
+class ADefAttack(Attack):
+    """Adversarial Attack that distorts the image, i.e. changes the locations
+    of pixels. The algorithm is described in [1],
```
Comment (jonasrauber) [inline foolbox/attacks/adef_attack.py]: there should be an underscore after [1] and [2]

File: foolbox/attacks/adef_attack.py [hunk 5]
```diff
@@ -0,0 +1,405 @@
...
+    max_norm : float
+        Maximum l2 norm of vector field (default max_norm = numpy.inf).
+    overshoot : float >= 1
+        Multiply the resulting vector field by this number,
+        if deformed image is still correctly classified
+        (default is overshoot = 1 for no overshooting).
+    smooth : float >= 0
+        Width of the Gaussian kernel used for smoothing.
+        (default is smooth = 0 for no smoothing).
+    targeting : bool
```
Comment (jonasrauber) [inline foolbox/attacks/adef_attack.py]: should be removed; use criterion instead

File: foolbox/attacks/adef_attack.py [hunk 6]
```diff
@@ -0,0 +1,405 @@
...
+        # Example 1: num_classes = 10 for MNIST
+        # Example 2: If ind_of_candidates == 1,
+        # then only the second highest label is targeted
+        num_classes = ind_of_candidates.size
+
+        n = 0  # iteration number
+
+        perturbed = a.original_image.copy()  # is updated in every iteration
+
+        image_original = a.original_image.copy()  # is not updated,
```
Comment (jonasrauber) [inline foolbox/attacks/adef_attack.py]: comments that are to long to fit after the code should go completely into a separate line before it

File: foolbox/attacks/adef_attack.py [hunk 7]
```diff
@@ -0,0 +1,405 @@
...
+        norm_full = 0  # norm of the vector field
+        vec_field_full = np.zeros((h, w, 2))  # the vector field
+
+        current_label = original_label
+        logging.info('Iterations finished: 0')
+        logging.info('Current labels: {} '.format(current_label))
+
+        for step in range(max_iter):
+            n += 1
+            logits, grad, is_adv = a.predictions_and_gradient(perturbed)
```
Comment (jonasrauber) [inline foolbox/attacks/adef_attack.py]: evaluating the gradient here seems unnecessary, now that backward is called later and the returned grad is never used as far as I see

File: foolbox/attacks/adef_attack.py [hunk 8]
```diff
@@ -0,0 +1,405 @@
...
+        # fx[lab] is negative if the model prefers the original label
+        # for x over the label 'lab'.
+        fx = (logits.transpose() - logits[0, original_label]).transpose()
+
+        norm_full = 0  # norm of the vector field
+        vec_field_full = np.zeros((h, w, 2))  # the vector field
+
+        current_label = original_label
+        logging.info('Iterations finished: 0')
+        logging.info('Current labels: {} '.format(current_label))
```
Comment (jonasrauber) [inline foolbox/attacks/adef_attack.py]: should be "label" not "labels" ?

File: foolbox/attacks/adef_attack.py [hunk 9]
```diff
@@ -0,0 +1,405 @@
...
+            grad_original = a.backward(logits_for_grad, perturbed)
+
+            # Find vector fields for the image and each candidate label.
+            # Keep the smallest vector field for each image.
+            norm_min = np.inf
+
+            # iterate over all candidate classes
+            for target_no in range(1, num_classes):
+
+                target_labels = candidates[0, target_no]
```
Comment (jonasrauber) [inline foolbox/attacks/adef_attack.py]: should be target_label, not target_labels ?!

File: foolbox/attacks/adef_attack.py [hunk 10]
```diff
@@ -0,0 +1,405 @@
...
+
+        color_axis = a.channel_axis(batch=False)  # get color axis
+        assert color_axis in [0, 2]
+        hw = [perturbed.shape[i] for i in range(perturbed.ndim)
+              if i != color_axis]
+        h, w = hw
+
+        logits, grad, is_adv = a.predictions_and_gradient(perturbed)
+
+        logits = np.expand_dims(logits, axis=0)
```
Comment (jonasrauber) [inline foolbox/attacks/adef_attack.py]: I don't understand why you expand the dimensions here. Wouldn't the following code be much simpler without this? Maybe there is a reason that I am overlooking right now; please explain (ideally also with a comment in the code) or simplify it if possible.

File: foolbox/attacks/adef_attack.py [hunk 11]
```diff
@@ -0,0 +1,405 @@
...
+                logits_for_grad = np.zeros_like(logits)
+                logits_for_grad[target_labels] = 1
+
+                # gradient of the target label w.r.t. image
+                grad_target = a.backward(logits_for_grad, perturbed)
+
+                # Derivative of the binary classifier 'F_lab - F_orig'
+                dfx = grad_target - grad_original
+
+                f_im = fx[0, candidates[0, target_no]]
```
Comment (jonasrauber) [inline foolbox/attacks/adef_attack.py]: `candidates[0, target_no]` can be replaced with `target_label`, no?

File: foolbox/attacks/adef_attack.py [hunk 12]
```diff
@@ -0,0 +1,405 @@
...
+
+                # create the vector field
+                vec_field_target = _create_vec_field(
+                    f_im, dfx, d1x, d2x, color_axis, smooth
+                )
```
Comment (jonasrauber) [inline foolbox/attacks/adef_attack.py]: closing bracket should be at the end of the previous line

File: foolbox/attacks/adef_attack.py [hunk 13]
```diff
@@ -0,0 +1,405 @@
...
+            # vector fields via vec_field_target += vec_field_full
+            perturbed = _compose(image_original.copy(), vec_field_min,
+                                 color_axis)
+
+            vec_field_full = vec_field_min
+            norm_full = norm_min
+
+            # getting the current label after applying the vector field
+            fx, _ = a.predictions(perturbed)
+            fx = np.expand_dims(fx, axis=0)
```
Comment (jonasrauber) [inline foolbox/attacks/adef_attack.py]: same as before: not sure why you expand the dimensions and then deal with transposes and 0-indices

File: foolbox/attacks/adef_attack.py [hunk 14]
```diff
@@ -0,0 +1,405 @@
...
+            current_label = np.argmax(fx, axis=1)
+            fx = (fx.transpose() - fx[0, current_label]).transpose()
+
+            # See if we have been successful.
```
Comment (jonasrauber) [inline foolbox/attacks/adef_attack.py]: Whether you have found a valid adversarial according to the criterion (be in targeted or untargeted) is something you get back when calling `a.predictions` in the form of the `is_adv` / `is_adversarial` return value. Along with getting rid of the `targeting` argument this code should be changed to to just use the return value of `a.predictions` rather than trying to determine by itself whether an adversarial has been found or not.

File: foolbox/attacks/adef_attack.py [hunk 15]
```diff
@@ -0,0 +1,405 @@
...
+                    'Image successfully deformed from {} to {}'.format(
+                        original_label, current_label))
+            logging.info('Iterations finished: {} '.format(n))
+            logging.info('Current labels: {} '.format(current_label))
+            logging.info('Norm vector field: {} '.format(norm_full))
+
+        # Overshooting
+        try_overshoot = False
+        have_highest_confidence = (fx == 0)
+        predicted_labels = np.where(have_highest_confidence)[0]
```
Comment (jonasrauber) [inline foolbox/attacks/adef_attack.py]: I've thought about this: I don't think we should handle the practically non-existent case of two classes with identical logits. We break the tie by defining the predicted class the one returned by np.argmax (it is deterministic).

File: foolbox/tests/test_attacks_adef_attack.py [hunk 16]
```diff
@@ -0,0 +1,51 @@
...
+    assert adv.image is not None
+    assert adv.distance.value < np.inf
+
+
+def test_targeted_attack_pytorch(bn_targeted_adversarial_pytorch):
+    adv = bn_targeted_adversarial_pytorch
+    attack = Attack()
+    attack(adv)
+    assert adv.image is None
+    assert adv.distance.value == np.inf
```
Comment (jonasrauber) [inline foolbox/tests/test_attacks_adef_attack.py]: this test should make sure that targeted attacks work, not that they fail

File: foolbox/tests/test_attacks_adef_attack.py [hunk 17]
```diff
@@ -0,0 +1,51 @@
...
+
+
+def test_targeted_attack(bn_targeted_adversarial):
+    adv = bn_targeted_adversarial
+    attack = Attack()
+    attack(adv)
+    assert adv.image is None
```
Comment (jonasrauber) [inline foolbox/tests/test_attacks_adef_attack.py]: this test should make sure that targeted attacks work, not that they fail

Comment (EvgeniaAR) [reply foolbox/tests/test_attacks_adef_attack.py]: alles klar, ich ändere das. ich habs von test_attacks_deepfool kopiert, line 38 :)

File: foolbox/attacks/adef_attack.py (hunk 10 above)
Comment (EvgeniaAR) [reply foolbox/attacks/adef_attack.py]: I will simplify it. I think the only reason was that the authors of the code did it and I compared with their code line after line. I did everything in the very same way to compare. I think they actually do need this other dimension for other features. In my case, I basically left this additional dimension blank. No further reasoning here, I will remove it :)

File: foolbox/attacks/adef_attack.py [hunk 18]
```diff
@@ -0,0 +1,372 @@
...
+    vec_field[:, :, 0] = -fval * alpha1 / norm_squared_alpha
+    vec_field[:, :, 1] = -fval * alpha2 / norm_squared_alpha
+
+    return vec_field
+
+
+class ADefAttack(Attack):
+    """Adversarial attack that distorts the image, i.e. changes the locations
+    of pixels. The algorithm is described in [1]_,
+    A Repository with the original code can be found in [2]_.
```
Comment (jonasrauber) [inline foolbox/attacks/adef_attack.py]: should still be lowercase ("a repository")

File: foolbox/attacks/adef_attack.py [hunk 19]
```diff
@@ -0,0 +1,372 @@
...
+        targeted = target_class is not None
+        original_label = a.original_class
+
+        # ADef targets classes according to their prediction score. If the
+        # attack is untargeted, ADef will take the labels of the top
+        # 'subsample' classes. This is faster than taking all the remaining
+        # 999 classes of ImageNet into account. For a targeted attack,
+        # it is necessary to find the probability of this class and pass
+        # this index to ind_of_candidates (not the actual target).
+        if targeted is False:
```
Comment (jonasrauber) [inline foolbox/attacks/adef_attack.py]: should be "if not targeted", or "if targeted" if you reverse if and else clause

File: foolbox/attacks/adef_attack.py [hunk 20]
```diff
@@ -0,0 +1,372 @@
...
+            # Include the original label in the list of targets (index 0).
+            logging.info('Only testing the top-{} classes'.format(subsample))
+            assert isinstance(subsample, int)
+            ind_of_candidates = np.arange(subsample)
+        else:
+            logits, _ = a.predictions(perturbed)
+            pred_sorted = (-logits).argsort()
+            index_of_target_class, = np.where(pred_sorted == target_class)
+            ind_of_candidates = index_of_target_class
+            # Include the original label in the list of targets (index 0).
```
Comment (jonasrauber) [inline foolbox/attacks/adef_attack.py]: as I said, you cannot assume that the original label is the one with the largest index

File: foolbox/attacks/adef_attack.py [hunk 21]
```diff
@@ -0,0 +1,372 @@
...
+            ind_of_candidates = np.unique(np.append(ind_of_candidates, 0))
+
+        # Remove negative entries.
```
Comment (jonasrauber) [inline foolbox/attacks/adef_attack.py]: why can there be negative indices?

File: foolbox/attacks/adef_attack.py (hunk 20 above)
Comment (EvgeniaAR) [reply foolbox/attacks/adef_attack.py]: ok, got rid of that altogether. It didn't seem to make much sense anyway. In the original code, they included the original label index in their list of targets, but then skipped it anyway when looping over possible targets and it seems never used it again. Hm.

**Comment by wielandbrendel** [comment]
@jonasrauber How is this PR going?

**Comment by jonasrauber** [comment]
Looks good, great work @EvgeniaAR.
Final things to do: add it to the docs, I guess in the `attacks.rst` file and the `attacks/gradient.rst` file, but compare to other attacks if that's all (not sure right now)
